tao_compiler/mlir/disc/transforms/disc_shape_optimization_utils.cc

```diff
@@ -0,0 +1,718 @@
+
+namespace mlir {
+namespace disc_ral {
+
+using ::mlir::func::FuncOp;
+
+// Gives a consistent order of a list op SymbolicDim Ops
+bool compareSymbolicDimOpNames(StringRef lhs, StringRef rhs) {
+  // S -> unknown dimension size at compile time
+  // C -> constant dimension size at compile time
+  // Not known encoding schema, fallback branch
+  if (lhs.size() < 1 || lhs[0] != 'S' && lhs[0] != 'C') return lhs < rhs;
+  if (rhs.size() < 1 || rhs[0] != 'S' && rhs[0] != 'C') return lhs < rhs;
+
+  int64_t lhsIdx, rhsIdx;
+  if (lhs.substr(1).getAsInteger(10, lhsIdx) ||
+      rhs.substr(1).getAsInteger(10, rhsIdx))
+    return lhs < rhs;
+  return (lhs[0] < rhs[0]) || (lhs[0] == rhs[0] && lhsIdx < rhsIdx);
+}
+
+// Gives a consistent order of a list op SymbolicDimProducts
+bool compareSymbolicDimProduct(const SymbolicDimProduct& lhs,
+                               const SymbolicDimProduct& rhs) {
+  if (lhs.symbols.size() < rhs.symbols.size()) return true;
+  if (lhs.symbols.size() == rhs.symbols.size()) {
+    for (const auto& z : llvm::zip(lhs.symbols, rhs.symbols)) {
+      StringRef lhsName = const_cast<SymbolicDimOp&>(std::get<0>(z)).getName();
+      StringRef rhsName = const_cast<SymbolicDimOp&>(std::get<1>(z)).getName();
+      if (compareSymbolicDimOpNames(lhsName, rhsName)) return true;
+      if (lhsName != rhsName) return false;
+    }
+  }
+  return false;
+}
+
+llvm::raw_ostream& operator<<(llvm::raw_ostream& os,
+                              const SymbolicDimProduct& product) {
+  os << "SymbolicDimProduct[\n\tfactor: " << product.factor << ",\n";
+  for (auto& s : product.symbols) os << "\tsymbol: " << s << "\n";
+  os << "]\n";
+  return os;
+}
+
+SymbolicDimMgr::SymbolicDimMgr(ModuleOp m) : m_(m) {}
+
+LogicalResult SymbolicDimMgr::load() {
+  m_.walk([&](disc_shape::SymbolicDimOp op) {
+    symbolDimUnionSet_[op] = op;
+    symbolNameSet_.insert(op.getName().str());
+  });
+  return loadShapeConstraintGraph();
+}
+
+std::string SymbolicDimMgr::getNextName() {
+  std::string name;
+  do {
+    name = (llvm::Twine("S") + llvm::Twine(nextSymbolicOpIdx_++)).str();
+  } while (!symbolNameSet_.insert(name).second);
+  return name;
+}
+
+SymbolicDimOp SymbolicDimMgr::newSymbolicDim() {
+  OpBuilder b(m_);
+  auto symbol = b.create<SymbolicDimOp>(m_.getLoc(), getNextName());
+  symbolDimUnionSet_[symbol] = symbol;
+  m_.push_back(symbol);
+  return symbol;
+}
+
+SymbolicDimOp SymbolicDimMgr::newConstantSymbolicDim(int64_t val) {
+  auto it = constantSymbolicDimMap_.find(val);
+  if (it == constantSymbolicDimMap_.end()) {
+    it = constantSymbolicDimMap_.insert(std::make_pair(val, newSymbolicDim()))
+             .first;
+    it->second.setDimSize(val);
+    it->second.setName((llvm::Twine("C") + llvm::Twine(val)).str());
+  }
+  return getRootSymbolicDim(it->second);
+}
+
+SymbolicDimOp SymbolicDimMgr::getRootSymbolicDim(SymbolicDimOp symbol) {
+  SymbolicDimOp current = symbol;
+  while (symbolDimUnionSet_[current] != current)
```

Comment:
How time-consuming is the recusive finding? Can the `rootSymbolicDim` be 'cached'?

Reply:
We actually update the union-find set during we visit the MLIR function, thus the cache is not stable. Furthermore, we merge each union-find group at the end of each iteration of stage 2 and then reduce the the number of unique SymbolicDimOp, which in turn reduce the overhead of this function in next iteration.